--- sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.hxx.old	2013-06-20 09:13:52.000000000 +0200
+++ sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.hxx	2013-07-16 15:49:35.000000000 +0200
@@ -178,7 +178,11 @@ namespace pdfi
         virtual void setDefaultCTM(double *ctm) SAL_OVERRIDE;
 
         // Start a page.
-        virtual void startPage(int pageNum, GfxState *state) SAL_OVERRIDE;
+        virtual void startPage(int pageNum, GfxState *state
+#if POPPLER_CHECK_VERSION(0, 23, 0)
+                               , XRef *xref
+#endif
+        ) SAL_OVERRIDE;
 
         // End a page.
         virtual void endPage() SAL_OVERRIDE;
--- sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx.old	2013-06-20 09:13:52.000000000 +0200
+++ sdext/source/pdfimport/xpdfwrapper/pdfioutdev_gpl.cxx	2013-07-16 15:49:21.000000000 +0200
@@ -491,7 +491,11 @@ PDFOutDev::~PDFOutDev()
     delete m_pUtf8Map;
 }
 
-void PDFOutDev::startPage(int /*pageNum*/, GfxState* state)
+void PDFOutDev::startPage(int /*pageNum*/, GfxState* state
+#if POPPLER_CHECK_VERSION(0, 23, 0)
+                          , XRef* /*xref*/
+#endif
+)
 {
     assert(state);
     printf("startPage %f %f\n",
