--- lib/SConscript.orig	2014-02-13 03:37:37.629916749 +0900
+++ lib/SConscript	2014-02-13 03:38:33.013782672 +0900
@@ -90,17 +90,13 @@
 else:
     # some distros use python2.5-config, others python-config2.5
     try:
-        env.ParseConfig(env['python_config'] + ' --cflags')
-        env.ParseConfig(env['python_config'] + ' --ldflags')
+        env.ParseConfig('pkg-config python-2.7 --cflags')
+        env.ParseConfig('pkg-config python-2.7 --libs')
     except OSError:
         print '%r does not work, trying python-config instead' % env['python_config']
         env.ParseConfig('python-config --ldflags')
         env.ParseConfig('python-config --cflags')
 
-if env.get('CPPDEFINES'):
-    # make sure assertions are enabled
-    env['CPPDEFINES'].remove('NDEBUG')
-
 # python extension module
 src = 'mypaintlib_wrap.cpp'
 module = build_py_module(env, '../_mypaintlib', Split(src), SHLIBPREFIX="")
