--- freemind/plugins/collaboration/socket/SocketBasics.java.orig	2013-07-12 12:37:54.507809109 +0900
+++ freemind/plugins/collaboration/socket/SocketBasics.java	2013-07-12 12:38:39.459807339 +0900
@@ -42,7 +42,7 @@
 import javax.swing.WindowConstants;
 
 import com.jgoodies.forms.builder.DefaultFormBuilder;
-import com.jgoodies.forms.factories.ButtonBarFactory;
+import com.jgoodies.forms.builder.ButtonBarBuilder;
 import com.jgoodies.forms.layout.FormLayout;
 
 import freemind.common.NumberProperty;
@@ -227,7 +227,7 @@
 			});
 			getRootPane().setDefaultButton(mOkButton);
 			getContentPane().add(
-					ButtonBarFactory.buildOKCancelBar(cancelButton, mOkButton),
+					new ButtonBarBuilder().addGlue().addButton(cancelButton, mOkButton).build(),
 					BorderLayout.SOUTH);
 			setTitle("Enter Password Dialog");
 			setDefaultCloseOperation(WindowConstants.DISPOSE_ON_CLOSE);
@@ -427,4 +427,4 @@
 	 */
 	public abstract void shutdown();
 
-}
\ No newline at end of file
+}
