--- freemind/freemind/preferences/layout/OptionPanel.java.orig	2013-07-12 12:26:34.227835908 +0900
+++ freemind/freemind/preferences/layout/OptionPanel.java	2013-07-12 12:28:32.003831268 +0900
@@ -49,7 +49,7 @@
 import javax.swing.UIManager.LookAndFeelInfo;
 
 import com.jgoodies.forms.builder.DefaultFormBuilder;
-import com.jgoodies.forms.factories.ButtonBarFactory;
+import com.jgoodies.forms.builder.ButtonBarBuilder;
 import com.jgoodies.forms.layout.FormLayout;
 import com.jgoodies.forms.layout.RowSpec;
 
@@ -238,7 +238,7 @@
 		});
 		frame.getRootPane().setDefaultButton(okButton);
 		frame.getContentPane().add(
-				ButtonBarFactory.buildOKCancelBar(cancelButton, okButton),
+				new ButtonBarBuilder().addGlue().addButton(cancelButton, okButton).build(),
 				BorderLayout.SOUTH);
 	}
 
