diff -uNr black-19.10b0.orig/_black_version.py black-19.10b0/_black_version.py
--- black-19.10b0.orig/_black_version.py	1970-01-01 09:00:00.000000000 +0900
+++ black-19.10b0/_black_version.py	2020-01-02 01:23:51.050207871 +0900
@@ -0,0 +1,2 @@
+version = "19.10b0"
+
diff -uNr black-19.10b0.orig/setup.cfg black-19.10b0/setup.cfg
--- black-19.10b0.orig/setup.cfg	2019-10-29 07:02:39.000000000 +0900
+++ black-19.10b0/setup.cfg	2020-01-02 01:22:31.008200161 +0900
@@ -1,5 +1,2 @@
 [bdist_wheel]
 python-tag = py36
-
-[options]
-setup_requires = setuptools_scm
diff -uNr black-19.10b0.orig/setup.py black-19.10b0/setup.py
--- black-19.10b0.orig/setup.py	2019-10-29 07:02:39.000000000 +0900
+++ black-19.10b0/setup.py	2020-01-02 01:24:20.197210679 +0900
@@ -17,10 +17,6 @@
 
 setup(
     name="black",
-    use_scm_version={
-        "write_to": "_black_version.py",
-        "write_to_template": 'version = "{version}"\n',
-    },
     description="The uncompromising code formatter.",
     long_description=get_long_description(),
     long_description_content_type="text/markdown",
