diff --git a/fontbox/src/main/java/org/apache/fontbox/cmap/CMap.java b/fontbox/src/main/java/org/apache/fontbox/cmap/CMap.java
index a9239df..b866a2d 100644
--- a/fontbox/src/main/java/org/apache/fontbox/cmap/CMap.java
+++ b/fontbox/src/main/java/org/apache/fontbox/cmap/CMap.java
@@ -164,14 +164,15 @@ public class CMap
      *
      * @return The CID that matches the lookup.
      */
-    public int lookupCID(byte[] code, int offset, int length) 
+    public int lookupCID(byte[] code, int offset, int length)
     {
         if (isInCodeSpaceRanges(code,offset,length)) 
         {
             int codeAsInt = getCodeFromArray(code, offset, length);
-            if (char2CIDMappings.containsKey(codeAsInt)) 
+            String codeAsStr = createStringFromBytes(code, offset, length);
+            if (char2CIDMappings.containsKey(codeAsStr)) 
             {
-                return char2CIDMappings.get(codeAsInt);
+                return char2CIDMappings.get(codeAsStr);
             } 
             else 
             {
@@ -188,6 +189,28 @@ public class CMap
         }
         return -1;
     }
+
+    private String createStringFromBytes(byte[] bytes, int offset, int length )
+    {
+        String retval = null;
+        try
+        {
+            if (length == 1)
+            {
+                retval = new String(bytes, offset, length, "ISO-8859-1");
+            }
+            else
+            {
+                retval = new String(bytes, offset, length, "UTF-16BE");
+            }
+        }
+        catch (IOException e)
+        {
+            // do nothing
+        }
+        return retval;
+    }
+
     
     /**
      * Convert the given part of a byte array to an integer.
