--- xml-commons-resolver-9999.orig/src/org/apache/xml/resolver/Catalog.java	2014-10-07 17:09:01.988059196 +0900
+++ xml-commons-resolver-9999/src/org/apache/xml/resolver/Catalog.java	2014-10-07 17:09:31.566090502 +0900
@@ -2243,13 +2243,15 @@
      */
     protected void addDelegate(CatalogEntry entry) {
         int pos = 0;
+        int type = entry.getEntryType();
         String partial = entry.getEntryArg(0);
 
         Enumeration local = localDelegate.elements();
         while (local.hasMoreElements()) {
             CatalogEntry dpe = (CatalogEntry) local.nextElement();
             String dp = dpe.getEntryArg(0);
-            if (dp.equals(partial)) {
+            int dptype = dpe.getEntryType();
+            if (dp.equals(partial) && (type == dptype)) {
                 // we already have this prefix
                 return;
             }
