--- xml-commons-resolver-1.2.orig/src/org/apache/xml/resolver/Catalog.java	2009-02-15 09:19:18.000000000 +0900
+++ xml-commons-resolver-1.2/src/org/apache/xml/resolver/Catalog.java	2009-02-15 09:22:02.000000000 +0900
@@ -31,6 +31,7 @@
 import java.util.Hashtable;
 import java.util.Locale;
 import java.util.Vector;
+import java.util.Iterator;
 
 import javax.xml.parsers.SAXParserFactory;
 
@@ -1165,6 +1166,23 @@
   }
 
 
+  /*
+   * Return all registered public IDs.
+   */
+  public Iterator getPublicIDs() {
+    Vector v = new Vector();
+    Enumeration e = catalogEntries.elements();
+
+    while (e.hasMoreElements()) {
+      CatalogEntry entry = (CatalogEntry) e.nextElement();
+      if (entry.getEntryType() == PUBLIC) {
+        v.add(entry.getEntryArg(0));
+      }
+    }
+    return v.iterator();
+  }
+
+
   /**
    * Return the applicable DOCTYPE system identifier.
    *
