--- xml-commons-resolver-1.2.orig/src/org/apache/xml/resolver/CatalogManager.java	2009-02-15 10:50:42.000000000 +0900
+++ xml-commons-resolver-1.2/src/org/apache/xml/resolver/CatalogManager.java	2009-02-15 10:54:33.000000000 +0900
@@ -207,7 +207,11 @@
     // setting, if defined.
   }
 
-  /** Constructor that specifies an explicit property file. */
+  /**
+   * Constructor that specifies an explicit property file.
+   * @param propertyFile path to property file (e.g. com/resources/CatalogManager).
+   * <code>null</code> means that no property file is consulted at all.
+   */
   public CatalogManager(String propertyFile) {
     this.propertyFile = propertyFile;
 
@@ -246,13 +250,14 @@
    * resources from it.
    */
   private synchronized void readProperties() {
+    if (propertyFile == null) return;
     try {
       propertyFileURI = CatalogManager.class.getResource("/"+propertyFile);
       InputStream in =
 	CatalogManager.class.getResourceAsStream("/"+propertyFile);
       if (in==null) {
 	if (!ignoreMissingProperties) {
-	  System.err.println("Cannot find "+propertyFile);
+	  debug.message(2, "Cannot find "+propertyFile);
 	  // there's no reason to give this warning more than once
 	  ignoreMissingProperties = true;
 	}
