diff --git a/nikola/plugins/compile/rest/__init__.py b/nikola/plugins/compile/rest/__init__.py
index e988f7b..855b968 100644
--- a/nikola/plugins/compile/rest/__init__.py
+++ b/nikola/plugins/compile/rest/__init__.py
@@ -35,7 +35,7 @@ try:
     import docutils.utils
     import docutils.io
     import docutils.readers.standalone
-    import docutils.writers.html4css1
+    import htmlwriter
     has_docutils = True
 except ImportError:
     has_docutils = False
@@ -96,7 +96,9 @@ class CompileRest(PageCompiler):
                         'syntax_highlight': 'short',
                         'math_output': 'mathjax',
                         'template': default_template_path,
-                    }, logger=self.logger, source_path=source, l_add_ln=add_ln, transforms=self.site.rst_transforms)
+                    }, logger=self.logger, source_path=source, l_add_ln=add_ln,
+                    transforms=self.site.rst_transforms,
+                    writer=htmlwriter.Writer())
                 if not isinstance(output, unicode_str):
                     # To prevent some weird bugs here or there.
                     # Original issue: empty files.  `output` became a bytestring.
