diff --git a/src/base/system_util.cc b/src/base/system_util.cc
index fcdb897f..3004aa3e 100644
--- a/src/base/system_util.cc
+++ b/src/base/system_util.cc
@@ -224,6 +224,7 @@ UserProfileDirectoryImpl::UserProfileDirectoryImpl() {
   return;
 #endif  // OS_ANDROID
 
+  const char *configuration_directory_env;
   std::string dir;
 
 #ifdef OS_IOS
@@ -237,8 +238,13 @@ UserProfileDirectoryImpl::UserProfileDirectoryImpl() {
 #endif  // OS_IOS
 
 #if defined(OS_WIN)
-  DCHECK(SUCCEEDED(Singleton<LocalAppDataDirectoryCache>::get()->result()));
-  dir = Singleton<LocalAppDataDirectoryCache>::get()->path();
+  configuration_directory_env = ::getenv("MOZC_CONFIGURATION_DIRECTORY");
+  if (configuration_directory_env) {
+    dir = configuration_directory_env;
+  } else {
+    DCHECK(SUCCEEDED(Singleton<LocalAppDataDirectoryCache>::get()->result()));
+    dir = Singleton<LocalAppDataDirectoryCache>::get()->path();
+  }
 # ifdef GOOGLE_JAPANESE_INPUT_BUILD
   dir = FileUtil::JoinPath(dir, kCompanyNameInEnglish);
   FileUtil::CreateDirectory(dir);
@@ -247,7 +253,12 @@ UserProfileDirectoryImpl::UserProfileDirectoryImpl() {
 #endif  // OS_WIN
 
 #if defined(__APPLE__)
-  dir = MacUtil::GetApplicationSupportDirectory();
+  configuration_directory_env = ::getenv("MOZC_CONFIGURATION_DIRECTORY");
+  if (configuration_directory_env) {
+    dir = configuration_directory_env;
+  } else {
+    dir = MacUtil::GetApplicationSupportDirectory();
+  }
 # ifdef GOOGLE_JAPANESE_INPUT_BUILD
   dir = FileUtil::JoinPath(dir, "Google");
   // The permission of ~/Library/Application Support/Google seems to be 0755.
@@ -260,14 +271,24 @@ UserProfileDirectoryImpl::UserProfileDirectoryImpl() {
 #endif  // __APPLE__
 
 #if defined(OS_LINUX)
-  char buf[1024];
-  struct passwd pw, *ppw;
-  const uid_t uid = geteuid();
-  CHECK_EQ(0, getpwuid_r(uid, &pw, buf, sizeof(buf), &ppw))
-      << "Can't get passwd entry for uid " << uid << ".";
-  CHECK_LT(0, strlen(pw.pw_dir))
-      << "Home directory for uid " << uid << " is not set.";
-  dir = FileUtil::JoinPath(pw.pw_dir, ".mozc");
+  configuration_directory_env = ::getenv("MOZC_CONFIGURATION_DIRECTORY");
+  if (configuration_directory_env) {
+    dir = configuration_directory_env;
+  } else {
+    const char *home_env = ::getenv("HOME");
+    if (home_env) {
+      dir = FileUtil::JoinPath(home_env, ".mozc");
+    } else {
+      char buf[1024];
+      struct passwd pw, *ppw;
+      const uid_t uid = geteuid();
+      CHECK_EQ(0, getpwuid_r(uid, &pw, buf, sizeof(buf), &ppw))
+          << "Can't get passwd entry for uid " << uid << ".";
+      CHECK_LT(0, strlen(pw.pw_dir))
+          << "Home directory for uid " << uid << " is not set.";
+      dir = FileUtil::JoinPath(pw.pw_dir, ".mozc");
+    }
+  }
 #endif  // OS_LINUX
 
   FileUtil::CreateDirectory(dir);
@@ -372,6 +393,10 @@ class ProgramFilesX86Cache {
 #endif  // OS_WIN
 
 std::string SystemUtil::GetServerDirectory() {
+  const char *server_directory_env = ::getenv("MOZC_SERVER_DIRECTORY");
+  if (server_directory_env) {
+    return server_directory_env;
+  }
 #ifdef OS_WIN
   DCHECK(SUCCEEDED(Singleton<ProgramFilesX86Cache>::get()->result()));
 # if defined(GOOGLE_JAPANESE_INPUT_BUILD)
@@ -430,6 +455,10 @@ std::string SystemUtil::GetToolPath() {
 }
 
 std::string SystemUtil::GetDocumentDirectory() {
+  const char *documents_directory_env = ::getenv("MOZC_DOCUMENTS_DIRECTORY");
+  if (documents_directory_env) {
+    return documents_directory_env;
+  }
 #if defined(__APPLE__)
   return GetServerDirectory();
 #elif defined(MOZC_DOCUMENT_DIRECTORY)
