--- requests-9999.orig/requests/compat.py	2014-08-07 23:38:23.695107794 +0900
+++ requests-9999/requests/compat.py	2014-08-07 23:44:55.082092376 +0900
@@ -4,7 +4,7 @@
 pythoncompat
 """
 
-from .packages import chardet
+import chardet
 
 import sys
 
--- requests-9999.orig/setup.py	2014-08-07 23:38:23.716107794 +0900
+++ requests-9999/setup.py	2014-08-07 23:44:43.404092836 +0900
@@ -17,7 +17,6 @@
 packages = [
     'requests',
     'requests.packages',
-    'requests.packages.chardet',
     'requests.packages.urllib3',
     'requests.packages.urllib3.packages',
     'requests.packages.urllib3.contrib',
