diff -uNr jarjar-9999.orig/src/main/com/tonicsystems/jarjar/KeepProcessor.java jarjar-9999/src/main/com/tonicsystems/jarjar/KeepProcessor.java
--- jarjar-9999.orig/src/main/com/tonicsystems/jarjar/KeepProcessor.java	2011-10-09 11:39:05.933960502 +0900
+++ jarjar-9999/src/main/com/tonicsystems/jarjar/KeepProcessor.java	2011-10-09 11:40:36.184832210 +0900
@@ -67,7 +67,7 @@
                     if (wildcard.matches(name))
                         roots.add(name);
                 depend.put(name, curSet = new HashSet<String>());
-                new ClassReader(new ByteArrayInputStream(struct.data)).accept(cv, 0);
+                new ClassReader(new ByteArrayInputStream(struct.data)).accept(cv, ClassReader.EXPAND_FRAMES);
                 curSet.remove(name);
             }
         } catch (Exception ignore) { }
