diff -uNr jdic-0.9.5-src.orig/jdic/demo/README jdic-0.9.5-src/jdic/demo/README
--- jdic-0.9.5-src.orig/jdic/demo/README	2008-03-13 23:09:13.000000000 +0900
+++ jdic-0.9.5-src/jdic/demo/README	2009-02-02 07:33:00.000000000 +0900
@@ -10,7 +10,7 @@
 
 Unix
 
-  java -classpath ../../jdic.jar:. -Djava.library.path=../../ <class name>
+  java -classpath /usr/share/jdic/lib/jdic.jar:. -Djava.library.path=/usr/lib <class name>
 
 where <class name> is the same as the application's name, such as FileExplorer,
 FileChooser, FileTypes, SimpleBrowser, Browser, and Tray. 
