diff -r 5b7cb45f4f6b src/blockdiag/utils/bootstrap.py
--- a/src/blockdiag/utils/bootstrap.py	Thu Sep 19 20:27:42 2013 +0900
+++ b/src/blockdiag/utils/bootstrap.py	Thu Sep 19 23:03:10 2013 +0900
@@ -37,9 +37,11 @@
         except SystemExit as e:
             return e
         except UnicodeEncodeError as e:
+            import traceback
+            traceback.print_exc()
             msg = "ERROR: UnicodeEncodeError caught " + \
                   "(check your font settings)\n"
-            sys.stderr.write(msg)
+            #sys.stderr.write(msg)
             return -1
         except Exception as e:
             if self.options and self.options.debug:
