--- ./hw/kdrive/linux/ps2.c.orig	2011-12-10 17:27:05.436476143 +0900
+++ ./hw/kdrive/linux/ps2.c	2011-12-10 17:27:46.813958855 +0900
@@ -29,6 +29,7 @@
 #include "inputstr.h"
 #include "scrnintstr.h"
 #include "kdrive.h"
+#include <stdint.h>
 
 static int
 Ps2ReadBytes (int fd, char *buf, int len, int min)
@@ -159,7 +160,7 @@
         return BadAlloc;
     }
 
-    pi->driverPrivate = (void *)fd;
+    pi->driverPrivate = (intptr_t)fd;
 
     return Success;
 }
@@ -168,7 +169,7 @@
 static void
 Ps2Disable (KdPointerInfo *pi)
 {
-    KdUnregisterFd (pi, (int)pi->driverPrivate, TRUE);
+    KdUnregisterFd (pi, (int)(intptr_t)pi->driverPrivate, TRUE);
 }
 
 static void
