--- ./hw/kdrive/linux/ms.c.orig	2011-12-10 17:24:42.374264669 +0900
+++ ./hw/kdrive/linux/ms.c	2011-12-10 17:26:22.588011823 +0900
@@ -32,6 +32,7 @@
 #include "inputstr.h"
 #include "scrnintstr.h"
 #include "kdrive.h"
+#include <stdint.h>
 
 static int
 MsReadBytes (int fd, char *buf, int len, int min)
@@ -152,7 +153,7 @@
     }
     if (KdRegisterFd (port, MsRead, pi))
 	return TRUE;
-    pi->driverPrivate = (void *)port;
+    pi->driverPrivate = (intptr_t)port;
 
     return Success;
 
@@ -164,7 +165,7 @@
 static void
 MsDisable (KdPointerInfo *pi)
 {
-    KdUnregisterFd (pi, (int)pi->driverPrivate, TRUE);
+    KdUnregisterFd (pi, (int)(intptr_t)pi->driverPrivate, TRUE);
 }
 
 static void
