diff -r b711ede115e3 src/reportlab/pdfgen/pdfimages.py
--- a/src/reportlab/pdfgen/pdfimages.py	Thu May 22 16:56:33 2014 +0100
+++ b/src/reportlab/pdfgen/pdfimages.py	Fri May 23 17:41:29 2014 +0900
@@ -122,7 +122,7 @@
         imagedata=['BI /W %d /H %d /BPC %d /CS /%s /F [%s/Fl] ID' % (imgwidth, imgheight, bpc, colorSpace, rl_config.useA85 and '/A85 ' or '')]
 
         #use a flate filter and, optionally, Ascii Base 85 to compress
-        raw = myimage.tostring()
+        raw = myimage.tobytes()
         rowstride = (imgwidth*bpc*bpp+7)>>3
         assert len(raw) == rowstride*imgheight, "Wrong amount of data for image"
         data = zlib.compress(raw)    #this bit is very fast...
