diff -r b711ede115e3 src/reportlab/pdfbase/ttfonts.py
--- a/src/reportlab/pdfbase/ttfonts.py	Thu May 22 16:56:33 2014 +0100
+++ b/src/reportlab/pdfbase/ttfonts.py	Fri May 23 17:17:32 2014 +0900
@@ -463,9 +463,9 @@
                     A = N.append
                     while length > 0:
                         char = self.read_ushort()
-                        A(chr(char))
+                        A(bytes([char]) if isPy3 else chr(char))
                         length -= 1
-                    N = ''.join(N)
+                    N = b''.join(N)
                 finally:
                     self._pos = opos
             elif platformId == 1 and encodingId == 0 and languageId == 0: # Macintosh, Roman, English, PS Name
