=== modified file 'src/main/java/org/apache/pdfbox/pdmodel/font/PDCIDFont.java'
--- src/main/java/org/apache/pdfbox/pdmodel/font/PDCIDFont.java	2009-03-26 21:45:40 +0000
+++ src/main/java/org/apache/pdfbox/pdmodel/font/PDCIDFont.java	2010-02-22 10:18:26 +0000
@@ -146,7 +146,7 @@
     public float getFontWidth( byte[] c, int offset, int length ) throws IOException
     {
 
-        float retval = 0.0f;
+    	float retval = getDefaultWidth();
         int code = getCodeFromArray( c, offset, length );
 
         Float widthFloat = (Float)widthCache.get( new Integer( code ) );

