diff --git a/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/font/PDType0TTFont.java b/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/font/PDType0TTFont.java
index 1c16887..783e454 100644
--- a/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/font/PDType0TTFont.java
+++ b/pdfbox/src/main/java/org/apache/pdfbox/pdmodel/font/PDType0TTFont.java
@@ -81,6 +81,9 @@ public class PDType0TTFont extends PDType0Font
     /** Map from unicode to cid */
     private CMAPEncodingEntry unicode2cidMap;
 
+    /** Use Unicode UCS-4 encoding for CMAP */
+    private boolean useUCS4 = false;
+
     /**
      * Constructor.
      *
@@ -178,11 +181,15 @@ public class PDType0TTFont extends PDType0Font
                 if (cmaps[i].getPlatformId() == CMAPTable.PLATFORM_WINDOWS)
                 {
                     int platformEncoding = cmaps[i].getPlatformEncodingId();
-                    if (CMAPTable.ENCODING_UNICODE == platformEncoding)
+                    if (useUCS4 && CMAPTable.ENCODING_UCS4 == platformEncoding)
                     {
                         gidMap = cmaps[i];
                         break;
                     }
+                    else if (CMAPTable.ENCODING_UNICODE == platformEncoding)
+                    {
+                        gidMap = cmaps[i];
+                    }
                 }
             }
 
@@ -351,6 +358,8 @@ public class PDType0TTFont extends PDType0Font
 
         CMAPTable cmapTable = ttf.getCMAP();
         CMAPEncodingEntry[] cmaps = cmapTable.getCmaps();
+        CMAPEncodingEntry cmap_bmp  = null;
+        CMAPEncodingEntry cmap_ucs4 = null;
 
         for (int i = 0; i < cmaps.length; i++)
         {
@@ -359,12 +368,26 @@ public class PDType0TTFont extends PDType0Font
                 int platformEncoding = cmaps[i].getPlatformEncodingId();
                 if (CMAPTable.ENCODING_UNICODE == platformEncoding)
                 {
-                    unicode2cidMap = cmaps[i];
+                    cmap_bmp = cmaps[i];
+                }
+                else if (CMAPTable.ENCODING_UCS4 == platformEncoding)
+                {
+                    cmap_ucs4 = cmaps[i];
                     break;
                 }
             }
         }
 
+        if (cmap_ucs4 != null)
+        {
+            unicode2cidMap = cmap_ucs4;
+            useUCS4 = true;
+        }
+        else
+        {
+            unicode2cidMap = cmap_bmp;
+        }
+        
         HorizontalMetricsTable hmtx = ttf.getHorizontalMetrics();
         int[] widths = hmtx.getAdvanceWidth();
         defaultW = Math.round(widths[0] * scaling);
@@ -475,5 +498,4 @@ public class PDType0TTFont extends PDType0Font
         return cidToGID.getCOSObject();
     }
 
-
 }
