Index: wicket/src/main/java/org/apache/wicket/protocol/http/WebApplication.java
===================================================================
--- wicket/src/main/java/org/apache/wicket/protocol/http/WebApplication.java	(revision 807334)
+++ wicket/src/main/java/org/apache/wicket/protocol/http/WebApplication.java	Mon Aug 24 21:00:23 CEST 2009
@@ -728,7 +728,8 @@
 				responsesPerSession = previousValue;
 			}
 		}
-		responsesPerSession.put(bufferId, renderedResponse);
+        String bufferKey = bufferId.startsWith("/") ? bufferId.substring(1) : bufferId;
+        responsesPerSession.put(bufferKey, renderedResponse);
 	}
 
 	/**
