--- mysql-connector-java-5.1.34.orig/src/testsuite/regression/DataSourceRegressionTest.java	2014-10-17 15:05:36.000000000 +0900
+++ mysql-connector-java-5.1.34/src/testsuite/regression/DataSourceRegressionTest.java	2015-03-01 21:11:23.123432457 +0900
@@ -58,7 +58,6 @@
 import com.mysql.jdbc.ConnectionProperties;
 import com.mysql.jdbc.MySQLConnection;
 import com.mysql.jdbc.NonRegisteringDriver;
-import com.mysql.jdbc.integration.jboss.MysqlValidConnectionChecker;
 import com.mysql.jdbc.jdbc2.optional.MysqlConnectionPoolDataSource;
 import com.mysql.jdbc.jdbc2.optional.MysqlDataSource;
 import com.mysql.jdbc.jdbc2.optional.MysqlDataSourceFactory;
@@ -313,29 +312,6 @@
         assertEquals(testIntFlag, thawedDs.getBlobSendChunkSize());
     }
 
-    /**
-     * Tests fix for BUG#20242 - MysqlValidConnectionChecker for JBoss doesn't
-     * work with MySQLXADataSources.
-     * 
-     * @throws Exception
-     *             if the test fails.
-     */
-    public void testBug20242() throws Exception {
-        if (versionMeetsMinimum(5, 0)) {
-            try {
-                Class.forName("org.jboss.resource.adapter.jdbc.ValidConnectionChecker");
-            } catch (Exception ex) {
-                return; // class not available for testing
-            }
-
-            MysqlXADataSource xaDs = new MysqlXADataSource();
-            xaDs.setUrl(dbUrl);
-
-            MysqlValidConnectionChecker checker = new MysqlValidConnectionChecker();
-            assertNull(checker.isValidConnection(xaDs.getXAConnection().getConnection()));
-        }
-    }
-
     private void bindDataSource(String name, DataSource ds) throws Exception {
         this.ctx.bind(this.tempDir.getAbsolutePath() + name, ds);
     }
