From: Ghislain Antony Vaillant <ghisvail@gmail.com>
Date: Wed, 11 Jan 2017 14:36:38 +0000
Subject: Fix CMake package version file

---
 CMakeLists.txt | 5 +++++
 1 file changed, 5 insertions(+)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -169,7 +169,12 @@
 	install(DIRECTORY glm DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 endif()
 
+# CMAKE_SIZEOF_VOID_P is temporarily unset to make the generated package
+# version file architecture-independent.
+set(_glm_CMAKE_SIZEOF_VOID_P ${CMAKE_SIZEOF_VOID_P})
+unset(CMAKE_SIZEOF_VOID_P)
 write_basic_package_version_file("${CMAKE_CURRENT_BINARY_DIR}/glmConfigVersion.cmake" VERSION ${GLM_VERSION} COMPATIBILITY AnyNewerVersion)
+set(CMAKE_SIZEOF_VOID_P ${_glm_CMAKE_SIZEOF_VOID_P})
 
 # build tree package config
 configure_file(cmake/glmBuildConfig.cmake.in glmConfig.cmake @ONLY)
