diff -uNr xf86-input-synaptics-9999.orig/test/fake-symbols.c xf86-input-synaptics-9999/test/fake-symbols.c
--- xf86-input-synaptics-9999.orig/test/fake-symbols.c	2011-10-09 20:32:58.084584527 +0900
+++ xf86-input-synaptics-9999/test/fake-symbols.c	2011-10-09 20:33:24.693251875 +0900
@@ -26,7 +26,7 @@
 }
 
 _X_EXPORT int
-xf86OpenSerial (pointer options)
+xf86OpenSerial (XF86OptionPtr options)
 {
     return 0;
 }
@@ -37,43 +37,43 @@
     return 0;
 }
 
-_X_EXPORT pointer
-xf86ReplaceIntOption(pointer optlist, const char *name, const int val)
+_X_EXPORT XF86OptionPtr
+xf86ReplaceIntOption(XF86OptionPtr optlist, const char *name, const int val)
 {
     return NULL;
 }
 
 _X_EXPORT char *
-xf86SetStrOption(pointer optlist, const char *name, char *deflt)
+xf86SetStrOption(XF86OptionPtr optlist, const char *name, char *deflt)
 {
     return NULL;
 }
 
 _X_EXPORT int
-xf86SetBoolOption(pointer optlist, const char *name, int deflt)
+xf86SetBoolOption(XF86OptionPtr optlist, const char *name, int deflt)
 {
     return 0;
 }
 
-_X_EXPORT pointer
-xf86AddNewOption(pointer head, const char *name, const char *val)
+_X_EXPORT XF86OptionPtr
+xf86AddNewOption(XF86OptionPtr head, const char *name, const char *val)
 {
     return NULL;
 }
 _X_EXPORT char *
-xf86FindOptionValue(pointer options, const char *name)
+xf86FindOptionValue(XF86OptionPtr options, const char *name)
 {
     return NULL;
 }
 
 _X_EXPORT char *
-xf86OptionName(pointer opt)
+xf86OptionName(XF86OptionPtr opt)
 {
     return NULL;
 }
 
 _X_EXPORT char *
-xf86OptionValue(pointer opt)
+xf86OptionValue(XF86OptionPtr opt)
 {
     return NULL;
 }
@@ -85,7 +85,7 @@
 }
 
 _X_EXPORT char *
-xf86CheckStrOption(pointer optlist, const char *name, char *deflt)
+xf86CheckStrOption(XF86OptionPtr optlist, const char *name, char *deflt)
 {
     return NULL;
 }
@@ -196,8 +196,8 @@
     return;
 }
 
-_X_EXPORT pointer
-xf86OptionListDuplicate(pointer options)
+_X_EXPORT XF86OptionPtr
+xf86OptionListDuplicate(XF86OptionPtr options)
 {
     return NULL;
 }
@@ -225,7 +225,7 @@
 }
 
 _X_EXPORT int
-xf86SetIntOption(pointer optlist, const char *name, int deflt)
+xf86SetIntOption(XF86OptionPtr optlist, const char *name, int deflt)
 {
     return 0;
 }
@@ -254,7 +254,7 @@
 _X_EXPORT int
 XIChangeDeviceProperty (DeviceIntPtr dev, Atom property, Atom type,
                         int format, int mode, unsigned long len,
-                        pointer value, Bool sendevent)
+                        XF86OptionPtr value, Bool sendevent)
 {
     return 0;
 }
@@ -367,15 +367,15 @@
 }
 
 
-_X_EXPORT pointer
-xf86ReplaceStrOption(pointer optlist, const char *name, const char* val)
+_X_EXPORT XF86OptionPtr
+xf86ReplaceStrOption(XF86OptionPtr optlist, const char *name, const char* val)
 {
     return NULL;
 }
 
 
-_X_EXPORT pointer
-xf86NextOption(pointer list)
+_X_EXPORT XF86OptionPtr
+xf86NextOption(XF86OptionPtr list)
 {
     return NULL;
 }
@@ -424,14 +424,14 @@
 
 #if GET_ABI_MAJOR(ABI_XINPUT_VERSION) < 12
 void
-xf86ProcessCommonOptions(InputInfoPtr pInfo, pointer list)
+xf86ProcessCommonOptions(InputInfoPtr pInfo, XF86OptionPtr list)
 {
 }
 
 void
 xf86CollectInputOptions(InputInfoPtr pInfo,
                         const char **defaultOpts,
-                        pointer extraOpts)
+                        XF86OptionPtr extraOpts)
 {
 }
 
diff -uNr xf86-input-synaptics-9999.orig/test/fake-symbols.h xf86-input-synaptics-9999/test/fake-symbols.h
--- xf86-input-synaptics-9999.orig/test/fake-symbols.h	2011-10-09 20:32:58.084584527 +0900
+++ xf86-input-synaptics-9999/test/fake-symbols.h	2011-10-09 20:33:24.693251875 +0900
@@ -5,25 +5,25 @@
 extern int xf86WriteSerial (int fd, const void *buf, int count);
 extern int xf86CloseSerial (int fd);
 extern int xf86WaitForInput (int fd, int timeout);
-extern int xf86OpenSerial (pointer options);
+extern int xf86OpenSerial (XF86OptionPtr options);
 extern int xf86SetSerialSpeed (int fd, int speed);
 
-extern pointer xf86ReplaceIntOption(pointer optlist, const char *name, const int val);
-extern pointer xf86AddNewOption(pointer head, const char *name, const char *val);
-extern char* xf86OptionName(pointer opt);
-extern char* xf86FindOptionValue(pointer options, const char *name);
+extern XF86OptionPtr xf86ReplaceIntOption(XF86OptionPtr optlist, const char *name, const int val);
+extern XF86OptionPtr xf86AddNewOption(XF86OptionPtr head, const char *name, const char *val);
+extern char* xf86OptionName(XF86OptionPtr opt);
+extern char* xf86FindOptionValue(XF86OptionPtr options, const char *name);
 extern int xf86NameCmp(const char *s1, const char *s2);
-extern char* xf86CheckStrOption(pointer optlist, const char *name, char *deflt);
+extern char* xf86CheckStrOption(XF86OptionPtr optlist, const char *name, char *deflt);
 
 
-extern char * xf86SetStrOption(pointer optlist, const char *name, char *deflt);
-extern int xf86SetBoolOption(pointer optlist, const char *name, int deflt);
-extern pointer xf86AddNewOption(pointer head, const char *name, const char *val);
-extern char* xf86FindOptionValue(pointer options, const char *name);
-extern char* xf86OptionName(pointer opt);
-extern char *xf86OptionValue(pointer opt);
+extern char * xf86SetStrOption(XF86OptionPtr optlist, const char *name, char *deflt);
+extern int xf86SetBoolOption(XF86OptionPtr optlist, const char *name, int deflt);
+extern XF86OptionPtr xf86AddNewOption(XF86OptionPtr head, const char *name, const char *val);
+extern char* xf86FindOptionValue(XF86OptionPtr options, const char *name);
+extern char* xf86OptionName(XF86OptionPtr opt);
+extern char *xf86OptionValue(XF86OptionPtr opt);
 extern int xf86NameCmp(const char *s1, const char *s2);
-extern char * xf86CheckStrOption(pointer optlist, const char *name, char *deflt);
+extern char * xf86CheckStrOption(XF86OptionPtr optlist, const char *name, char *deflt);
 extern void xf86AddEnabledDevice(InputInfoPtr pInfo);
 extern void xf86RemoveEnabledDevice(InputInfoPtr pInfo);
 extern Atom XIGetKnownProperty(char *name);
@@ -53,7 +53,7 @@
 
 extern InputInfoPtr xf86FirstLocalDevice(void);
 extern void xf86DeleteInput(InputInfoPtr pInp, int flags);
-extern pointer xf86OptionListDuplicate(pointer options);
+extern XF86OptionPtr xf86OptionListDuplicate(XF86OptionPtr options);
 extern Bool
 InitButtonClassDeviceStruct(DeviceIntPtr dev, int numButtons, Atom* labels,
                             CARD8 *map);
@@ -65,7 +65,7 @@
                       unsigned int      key_code,
                       int               is_down);
 extern int
-xf86SetIntOption(pointer optlist, const char *name, int deflt);
+xf86SetIntOption(XF86OptionPtr optlist, const char *name, int deflt);
 extern void
 xf86PostButtonEventP(DeviceIntPtr	device,
                      int		is_absolute,
@@ -83,7 +83,7 @@
 extern int
 XIChangeDeviceProperty (DeviceIntPtr dev, Atom property, Atom type,
                         int format, int mode, unsigned long len,
-                        pointer value, Bool sendevent);
+                        XF86OptionPtr value, Bool sendevent);
 extern CARD32 GetTimeInMillis (void);
 
 extern int
@@ -132,10 +132,10 @@
 InitValuatorClassDeviceStruct(DeviceIntPtr dev, int numAxes, Atom *labels,
                               int numMotionEvents, int mode);
 
-extern pointer
-xf86ReplaceStrOption(pointer optlist, const char *name, const char* val);
+extern XF86OptionPtr
+xf86ReplaceStrOption(XF86OptionPtr optlist, const char *name, const char* val);
 
-extern pointer xf86NextOption(pointer list);
+extern XF86OptionPtr xf86NextOption(XF86OptionPtr list);
 
 extern int
 XIGetDeviceProperty (DeviceIntPtr dev, Atom property, XIPropertyValuePtr *value);
@@ -159,12 +159,12 @@
 
 #if GET_ABI_MAJOR(ABI_XINPUT_VERSION) < 12
 extern void
-xf86ProcessCommonOptions(InputInfoPtr pInfo, pointer list);
+xf86ProcessCommonOptions(InputInfoPtr pInfo, XF86OptionPtr list);
 
 extern void
 xf86CollectInputOptions(InputInfoPtr pInfo,
                         const char **defaultOpts,
-                        pointer extraOpts);
+                        XF86OptionPtr extraOpts);
 
 extern InputInfoPtr
 xf86AllocateInput(InputDriverPtr drv, int flags);
@@ -173,6 +173,6 @@
 extern ClientPtr serverClient;
 
 extern Bool QueueWorkProc (
-    Bool (*function)(ClientPtr /* pClient */, pointer /* closure */),
-    ClientPtr client, pointer closure);
+    Bool (*function)(ClientPtr /* pClient */, XF86OptionPtr /* closure */),
+    ClientPtr client, XF86OptionPtr closure);
 #endif
