--- popup.el.orig	2012-04-29 02:50:25.617716759 +0900
+++ popup.el	2012-04-29 02:54:48.561429508 +0900
@@ -143,7 +143,7 @@
   (loop with tab-width = 4
         for item in list
         for summary = (popup-item-summary item)
-        maximize (string-width (popup-x-to-string item)) into width
+        maximize (string-width (popup-x-to-string (or (get-text-property 0 'view item) item))) into width
         if (stringp summary)
         maximize (+ (string-width summary) 2) into summary-width
         finally return
@@ -409,10 +409,10 @@
                                (+ summary-width 2)
                              0)))
          (string (car (popup-substring-by-width string content-width)))
-         (string-width (string-width string))
+         (string-width (string-width (or (get-text-property 0 'view string) string)))
          (spacing (max (- popup-width string-width summary-width) 0)))
     (concat margin-left
-            string
+            (or (get-text-property 0 'view string) string)
             (make-string spacing ? )
             summary
             symbol
