diff --git a/control.c b/control.c
index f629bfb..70a8df7 100644
--- a/control.c
+++ b/control.c
@@ -126,7 +126,7 @@ void hello (void *tun)
     struct buffer *buf;
     struct tunnel *t;
     struct timeval tv;
-    tv.tv_sec = HELLO_DELAY;
+    tv.tv_sec = gconfig.hellodelay ? gconfig.hellodelay : HELLO_DELAY;
     tv.tv_usec = 0;
     t = (struct tunnel *) tun;
     buf = new_outgoing (t);
diff --git a/file.c b/file.c
index 9a8cfd6..a10ce4c 100644
--- a/file.c
+++ b/file.c
@@ -47,6 +47,9 @@ int init_config ()
     gconfig.packet_dump = 0;
     gconfig.debug_tunnel = 0;
     gconfig.debug_state = 0;
+    gconfig.hellodelay = 0;
+    gconfig.ctrl_retrans_max = 0;
+    gconfig.ctrl_retrans_delay = 0;
     lnslist = NULL;
     laclist = NULL;
     deflac = (struct lac *) calloc (1, sizeof (struct lac));
@@ -1208,6 +1211,65 @@ int set_rand_source (char *word, char *value, int context, void *item)
     }
 }
 
+int set_hellodelay (char *word, char *value, int context, void *item)
+{
+    switch (context & ~CONTEXT_DEFAULT)
+    {
+    case CONTEXT_GLOBAL:
+#ifdef DEBUG_FILE
+        l2tp_log (LOG_DEBUG, "set_hellodelay: Setting HELLO delay %s\n",
+            value);
+#endif
+        set_int (word, value, &(((struct global *) item)->hellodelay));
+        break;
+    default:
+        snprintf (filerr, sizeof (filerr), "'%s' not valid in this context\n",
+            word);
+        return -1;
+    }
+    return 0;
+}
+
+int set_ctrl_retrans_max (char *word, char *value, int context, void *item)
+{
+    switch (context & ~CONTEXT_DEFAULT)
+    {
+    case CONTEXT_GLOBAL:
+#ifdef DEBUG_FILE
+        l2tp_log (LOG_DEBUG,
+            "set_ctrl_retrans_max: Setting Control packet retrans max %s\n",
+            value);
+#endif
+        set_int (word, value, &(((struct global *) item)->ctrl_retrans_max));
+        break;
+    default:
+        snprintf (filerr, sizeof (filerr), "'%s' not valid in this context\n",
+            word);
+        return -1;
+    }
+    return 0;
+}
+
+int set_ctrl_retrans_delay (char *word, char *value, int context, void *item)
+{
+    switch (context & ~CONTEXT_DEFAULT)
+    {
+    case CONTEXT_GLOBAL:
+#ifdef DEBUG_FILE
+        l2tp_log (LOG_DEBUG,
+            "set_ctrl_retrans_delay: Setting Control packet retrans delay %s\n",
+            value);
+#endif
+        set_int (word, value, &(((struct global *) item)->ctrl_retrans_delay));
+        break;
+    default:
+        snprintf (filerr, sizeof (filerr), "'%s' not valid in this context\n",
+            word);
+        return -1;
+    }
+    return 0;
+}
+
 int parse_config (FILE * f)
 {
     /* Read in the configuration file handed to us */
@@ -1485,5 +1547,8 @@ struct keyword words[] = {
     {"tx bps", &set_speed},
     {"rx bps", &set_speed},
     {"bps", &set_speed},
+    {"hello delay", &set_hellodelay},
+    {"ctrl retrans max", &set_ctrl_retrans_max},
+    {"ctrl retrans delay", &set_ctrl_retrans_delay},
     {NULL, NULL}
 };
diff --git a/file.h b/file.h
index e54aa61..224f5e2 100644
--- a/file.h
+++ b/file.h
@@ -157,6 +157,9 @@ struct global
     int debug_tunnel;		/* Print tunnel debugging info? */
     int debug_state;		/* Print FSM debugging info? */
     int ipsecsaref;
+    int hellodelay;		/* HELLO packet interval */
+    int ctrl_retrans_max;	/* Control packet retrans max count */
+    int ctrl_retrans_delay;		/*Control packet retrans delay */
 };
 
 extern struct global gconfig;   /* Global configuration options */
diff --git a/network.c b/network.c
index f5e835b..9117745 100644
--- a/network.c
+++ b/network.c
@@ -186,6 +186,8 @@ void control_xmit (void *b)
     struct tunnel *t;
     struct timeval tv;
     int ns;
+    int retrans_max = gconfig.ctrl_retrans_max ? gconfig.ctrl_retrans_max
+	                                           : DEFAULT_MAX_RETRIES;
 
     if (!buf)
     {
@@ -216,7 +218,7 @@ void control_xmit (void *b)
             return;
         }
     }
-    if (buf->retries > DEFAULT_MAX_RETRIES)
+    if (buf->retries > retrans_max)
     {
         /*
            * Too many retries.  Either kill the tunnel, or
@@ -249,7 +251,7 @@ void control_xmit (void *b)
         /*
            * FIXME:  How about adaptive timeouts?
          */
-        tv.tv_sec = 1;
+        tv.tv_sec = gconfig.ctrl_retrans_delay ? gconfig.ctrl_retrans_delay : 1;
         tv.tv_usec = 0;
         schedule (tv, control_xmit, buf);
 #ifdef DEBUG_CONTROL_XMIT
