diff --git a/xl2tpd.c b/xl2tpd.c
index e0955af..af15f2e 100644
--- a/xl2tpd.c
+++ b/xl2tpd.c
@@ -1489,8 +1489,27 @@ static void consider_pidfile() {
     }
 }
 
+/* from tpm_emulator tpmd.c */
+static int mkdirs(const char *path)
+{
+    char *copy = strdup(path);
+    char *p = strchr(copy + 1, '/');
+    while (p != NULL) {
+        *p = '\0';
+        if ((mkdir(copy, 0755) == -1) && (errno != EEXIST)) {
+            free(copy);
+            return errno;
+        }
+        *p = '/';
+        p = strchr(p + 1, '/');
+    }
+    free(copy);
+    return 0;
+}
+
 static void open_controlfd() 
 {
+    mkdirs(gconfig.controlfile);
     control_fd = open (gconfig.controlfile, O_RDONLY | O_NONBLOCK, 0600);
     if (control_fd < 0)
     {
