--- xl2tpd-1.2.8.orig/Makefile	2011-02-23 14:10:21.000000000 +0900
+++ xl2tpd-1.2.8/Makefile	2011-07-12 09:55:41.253710445 +0900
@@ -53,7 +53,7 @@
 # but we use a local copy if we don't find it.
 #
 #KERNELSRC=/lib/modules/`uname -r`/build/
-KERNELSRC?=./linux
+KERNELSRC?=/usr/src/linux
 OSFLAGS?= -DLINUX -I$(KERNELSRC)/include/
 #
 # Uncomment the following to use the kernel interface under Linux
@@ -62,7 +62,7 @@
 # are packages seperately (eg kernel-headers on Fedora)
 # Note: 2.6.23+ support still needs some changes in the xl2tpd source
 #
-#OSFLAGS+= -DUSE_KERNEL
+OSFLAGS+= -DUSE_KERNEL
 #
 #
 # Uncomment the next line for FreeBSD
