--- rr_lib.rb.orig	2013-08-23 01:35:30.613734829 +0900
+++ rr_lib.rb	2013-08-23 01:36:03.340733540 +0900
@@ -1858,7 +1858,7 @@
 		is #{@settings['cd'].getFileSize(track)} bytes." if @settings['debug'] 
 
 		if installed('df')				
-			freeDiskSpace = `LANG=C df \"#{@settings['Out'].getDir()}\"`.split()[10].to_i
+			freeDiskSpace = `LC_ALL=C df \"#{@settings['Out'].getDir()}\"`.split()[10].to_i
 			puts "Free disk space is #{freeDiskSpace} MB" if @settings['debug']
 			if @settings['cd'].getFileSize(track) > freeDiskSpace*1000
 				@settings['log'].add(_("Not enough disk space left! Rip aborted"))
