--- jellytools.platform/src/org/netbeans/jellytools/TreeTableOperator.java.orig	2012-04-01 20:24:45.178270610 +0900
+++ jellytools.platform/src/org/netbeans/jellytools/TreeTableOperator.java	2012-04-01 20:25:01.579065572 +0900
@@ -294,6 +294,13 @@
             ComponentOperator realOper = ((RenderedTreeOperator)oper).getRealOperator();
             DriverManager.getMouseDriver(realOper).exitMouse(realOper);
         }
+
+        @org.netbeans.api.annotations.common.SuppressWarnings("BC_UNCONFIRMED_CAST") //suppres FindBugs warning
+        public void wheelMouse(ComponentOperator oper, int x, int y, int scrollAmount) {
+            checkSupported(oper);
+            ComponentOperator realOper = ((RenderedTreeOperator)oper).getRealOperator();
+            DriverManager.getMouseDriver(realOper).wheelMouse(realOper, x, y, scrollAmount);
+        }
     }
     
     /** Operator used to process events not on JTree but on TreeTable
