--- java/net/Socket.java.orig	2015-03-03 16:15:55.814299837 +0900
+++ java/net/Socket.java	2015-03-03 16:18:32.861784192 +0900
@@ -274,6 +274,11 @@
   }
 
   /**
+   * dummy
+   */
+  public Socket(java.net.Proxy proxy) {}
+
+  /**
    * This constructor is where the real work takes place.  Connect to the
    * specified address and port.  Use default local values if not specified,
    * otherwise use the local host and port passed in.  Create as stream or
