--- Makefile.am.orig	2011-12-16 22:14:11.080015785 +0900
+++ Makefile.am	2011-12-16 22:15:21.209139050 +0900
@@ -1276,6 +1276,7 @@
 	ln -sf $(XALAN2_SERIALIZER_JAR) \
 	  $(ECJ_BOOT_DIR)/lib/endorsed/xalan-j2-serializer.jar && \
 	ln -sf $(XERCES2_JAR) $(ECJ_BOOT_DIR)/lib/endorsed/xerces-j2.jar
+	ln -sf $(XMLCOMMONSEXTERNAL_JAR) $(ECJ_BOOT_DIR)/lib/endorsed/xml-apis.jar
 	mkdir -p $(ECJ_BOOT_DIR)/jre/lib && \
 	cp $(SYSTEM_JDK_DIR)/jre/lib/rt.jar $(ECJ_BOOT_DIR)/jre/lib/rt.jar && \
 	ln -sf $(SYSTEM_JDK_DIR)/jre/lib/$(JRE_ARCH_DIR) \
