--- acinclude.m4.orig	2011-12-16 22:04:50.385025437 +0900
+++ acinclude.m4	2011-12-16 22:22:13.156988994 +0900
@@ -556,16 +556,18 @@
 [
   AC_MSG_CHECKING(for endorsed jars dir)
   AC_ARG_WITH([endorsed-dir],
-              [AS_HELP_STRING(--with-endorsed-dir,specify directory of endorsed jars (xalan-j2.jar, xalan-j2-serializer.jar, xerces-j2.jar))],
+              [AS_HELP_STRING(--with-endorsed-dir,specify directory of endorsed jars (xalan.jar, serializer.jar, xerces.jar))],
   [
     if test "x${withval}" = "xno"; then
         ENDORSED_JARS="${withval}"
         AC_MSG_RESULT(${withval})
-    else if test -f "${withval}/xalan-j2.jar"; then
-      if test -f "${withval}/xalan-j2-serializer.jar"; then
-        if test -f "${withval}/xerces-j2.jar"; then
-          ENDORSED_JARS="${withval}"
-          AC_MSG_RESULT(${withval})
+    else if test -f "${withval}/xalan.jar"; then
+      if test -f "${withval}/serializer.jar"; then
+        if test -f "${withval}/xercesImpl.jar"; then
+          if test -f "${withval}/xml-apis.jar"; then
+            ENDORSED_JARS="${withval}"
+            AC_MSG_RESULT(${withval})
+          fi
         fi
       fi
     fi
@@ -574,10 +576,11 @@
     ENDORSED_JARS=
   ])
   if test -z "${ENDORSED_JARS}"; then
-    if test -f "/usr/share/java/xalan-j2.jar"; then
-      if test -f "/usr/share/java/xalan-j2-serializer.jar"; then
-        if test -f "/usr/share/java/xerces-j2.jar"; then
-          ENDORSED_JARS="/usr/share/java/xalan-j2.jar /usr/share/java/xalan-j2-serializer.jar /usr/share/java/xerces-j2.jar"
+    if test -f "/usr/share/xalan/lib/xalan.jar"; then
+      if test -f "/usr/share/xalan/lib/serializer.jar"; then
+        if test -f "/usr/share/xerces-2/lib/xercesImpl.jar"; then
+          if test -f "/usr/share/xml-commons-1.4/lib/xml-apis.jar"
+          ENDORSED_JARS="/usr/share/xalan/lib/xalan.jar /usr/share/xalan/lib/serializer.jar /usr/share/xerces-2/lib/xercesImpl.jar /usr/share/xml-commons-external-1.4/lib/xml-apis.jar"
           AC_MSG_RESULT(/usr/share/java)
         fi
       fi
@@ -587,7 +590,7 @@
     fi
   fi
   if test -z "${ENDORSED_JARS}"; then
-    AC_MSG_ERROR("A directory containing required jars (xalan-j2.jar, xalan-j2-serializer.jar, xerces-j2.jar) was not found.")
+    AC_MSG_ERROR("A directory containing required jars (xalan.jar, serializer.jar, xerces.jar) was not found.")
   fi
   AC_SUBST(ENDORSED_JARS)
 ])
@@ -644,10 +647,10 @@
     XALAN2_JAR=
   ])
   if test -z "${XALAN2_JAR}"; then
-    if test -e "/usr/share/java/xalan-j2.jar"; then
-      XALAN2_JAR=/usr/share/java/xalan-j2.jar
-    elif test -e "/usr/share/java/xalan2.jar"; then
-      XALAN2_JAR=/usr/share/java/xalan2.jar
+    if test -e "/usr/share/xalan/lib/xalan.jar"; then
+      XALAN2_JAR=/usr/share/xalan/lib/xalan.jar
+    elif test -e "/usr/share/xalan/lib/xalan.jar"; then
+      XALAN2_JAR=/usr/share/xalan/lib/xalan.jar
     elif test -e "/usr/share/xalan/lib/xalan.jar"; then
       XALAN2_JAR=/usr/share/xalan/lib/xalan.jar
     else
@@ -655,7 +658,7 @@
     fi
   fi
   if test -z "${XALAN2_JAR}"; then
-    AC_MSG_ERROR("A xalan2 jar was not found.")
+    AC_MSG_ERROR("A xalan jar was not found.")
   fi
   AC_MSG_RESULT(${XALAN2_JAR})
   AC_SUBST(XALAN2_JAR)
@@ -675,12 +678,10 @@
     XALAN2_SERIALIZER_JAR=
   ])
   if test -z "${XALAN2_SERIALIZER_JAR}"; then
-    if test -e "/usr/share/java/xalan-j2-serializer.jar"; then
-      XALAN2_SERIALIZER_JAR=/usr/share/java/xalan-j2-serializer.jar
+    if test -e "/usr/share/xalan/lib/serializer.jar"; then
+      XALAN2_SERIALIZER_JAR=/usr/share/xalan/lib/serializer.jar
     elif test -e "/usr/share/xalan-serializer/lib/serializer.jar"; then
       XALAN2_SERIALIZER_JAR=/usr/share/xalan-serializer/lib/serializer.jar
-    elif test -e "/usr/share/java/serializer.jar"; then
-      XALAN2_SERIALIZER_JAR=/usr/share/java/serializer.jar
     else
       AC_MSG_RESULT(no)
     fi
@@ -706,8 +707,8 @@
     XERCES2_JAR=
   ])
   if test -z "${XERCES2_JAR}"; then
-    if test -e "/usr/share/java/xerces-j2.jar"; then
-      XERCES2_JAR=/usr/share/java/xerces-j2.jar
+    if test -e "/usr/share/xerces-2/lib/xercesImpl.jar"; then
+      XERCES2_JAR=/usr/share/xerces-2/lib/xercesImpl.jar
     elif test -e "/usr/share/java/xerces2.jar"; then
       XERCES2_JAR=/usr/share/java/xerces2.jar
     elif test -e "/usr/share/xerces-2/lib/xercesImpl.jar"; then
@@ -725,6 +726,33 @@
   AC_SUBST(XERCES2_JAR)
 ])
 
+AC_DEFUN([IT_FIND_XMLCOMMONSEXTERNAL_JAR],
+[
+  AC_MSG_CHECKING([for a xml-commons-external jar])
+  AC_ARG_WITH([xml-commons-external-jar],
+              [AS_HELP_STRING(--with-xml-commons-external-jar,specify location of the xml-commons-external jar)],
+  [
+    if test -f "${withval}" ; then
+      XMLCOMMONSEXTERNAL_JAR="${withval}"
+    fi
+  ],
+  [
+    XMLCOMMONSEXTERNAL_JAR=
+  ])
+  if test -z "${XMLCOMMONSEXTERNAL_JAR}"; then
+    if test -e "/usr/share/xml-commons-external-1.4/lib/xml-apis.jar"; then
+      XMLCOMMONSEXTERNAL_JAR=/usr/share/xml-commons-external-1.4/lib/xml-apis.jar
+    else
+      AC_MSG_RESULT(no)
+    fi
+  fi
+  if test -z "${XMLCOMMONSEXTERNAL_JAR}"; then
+    AC_MSG_ERROR("A xml-commons-external jar was not found.")
+  fi
+  AC_MSG_RESULT(${XMLCOMMONSEXTERNAL_JAR})
+  AC_SUBST(XMLCOMMONSEXTERNAL_JAR)
+])
+
 AC_DEFUN([IT_FIND_RHINO_JAR],
 [
   AC_MSG_CHECKING([whether to include Javascript support via Rhino])
