--- Makefile.in.orig	2011-12-16 22:15:33.700982880 +0900
+++ Makefile.in	2011-12-16 22:20:10.045528099 +0900
@@ -294,6 +294,7 @@
 XALAN2_JAR = @XALAN2_JAR@
 XALAN2_SERIALIZER_JAR = @XALAN2_SERIALIZER_JAR@
 XERCES2_JAR = @XERCES2_JAR@
+XMLCOMMONSEXTERNAL_JAR = @XMLCOMMONSEXTERNAL_JAR@
 XINERAMA_CFLAGS = @XINERAMA_CFLAGS@
 XINERAMA_LIBS = @XINERAMA_LIBS@
 XPROTO_CFLAGS = @XPROTO_CFLAGS@
@@ -1713,6 +1714,7 @@
 	ln -sf $(XALAN2_SERIALIZER_JAR) \
 	  $(ECJ_BOOT_DIR)/lib/endorsed/xalan-j2-serializer.jar && \
 	ln -sf $(XERCES2_JAR) $(ECJ_BOOT_DIR)/lib/endorsed/xerces-j2.jar
+	ln -sf $(XMLCOMMONSEXTERNAL_JAR) $(ECJ_BOOT_DIR)/lib/endorsed/xml-apis.jar
 	mkdir -p $(ECJ_BOOT_DIR)/jre/lib && \
 	cp $(SYSTEM_JDK_DIR)/jre/lib/rt.jar $(ECJ_BOOT_DIR)/jre/lib/rt.jar && \
 	ln -sf $(SYSTEM_JDK_DIR)/jre/lib/$(JRE_ARCH_DIR) \
